map_creator/src/find_base_byPCA.cpp

```diff
@@ -0,0 +1,334 @@
+    base_trns.setOrigin(base_vec);
+    base_trns.setRotation(base_quat);
+
+//Inverse of the transformation
+    tf2::Transform base_trns_inv;
+    base_trns_inv = base_trns.inverse();
+
+//Creating a transformation of grasp pose
+    tf2::Vector3 grasp_vec(grasp_pose.position.x, grasp_pose.position.y, grasp_pose.position.z);
+    tf2::Quaternion grasp_quat(grasp_pose.orientation.x, grasp_pose.orientation.y, grasp_pose.orientation.z, grasp_pose.orientation.w);
+    grasp_quat.normalize();
+    tf2::Transform grasp_trns;
+    grasp_trns.setOrigin(grasp_vec);
+    grasp_trns.setRotation(grasp_quat);
+
+//Transforming grasp pose to origin from where we can check for Ik
+    tf2::Transform new_grasp_trns;
+    //new_grasp_trns = grasp_trns * base_trns_inv;
+    new_grasp_trns =  base_trns_inv * grasp_trns;
+//Creating a new grasp pose in the origin co-ordinate
+    vector<double> new_grasp_pos;
+    tf2::Vector3 new_grasp_vec;
+    tf2::Quaternion new_grasp_quat;
+    new_grasp_vec =  new_grasp_trns.getOrigin();
+    new_grasp_quat =  new_grasp_trns.getRotation();
+    new_grasp_quat.normalize();
+    new_grasp_pos.push_back(new_grasp_vec[0]);
+    new_grasp_pos.push_back(new_grasp_vec[1]);
+    new_grasp_pos.push_back(new_grasp_vec[2]);
+    new_grasp_pos.push_back(new_grasp_quat[0]);
+    new_grasp_pos.push_back(new_grasp_quat[1]);
+    new_grasp_pos.push_back(new_grasp_quat[2]);
+    new_grasp_pos.push_back(new_grasp_quat[3]);
+
+//Check the new grasp_pose for Ik
+    Kinematics k;
+    std::vector<double> joints;
+    
+    joints.resize(6);
+    if(k.isIKSuccess(new_grasp_pos, joints, numOfSolns))
+      return true;
+    else
+      return false;
+
+} 
+
+
+
+void poseToEigenVector(const geometry_msgs::Pose pose, Eigen::VectorXd& vec)
+{
+    vec<<pose.position.x, pose.position.y, pose.position.z, pose.orientation.x, pose.orientation.y, pose.orientation.z, pose.orientation.w;
+    
+}
+
+
+
+void findOptimalPosebyPCA(const vector<geometry_msgs::Pose> probBasePoses,geometry_msgs::Pose& final_base_pose){
+ 
+    Eigen::MatrixXd poseData( 7,probBasePoses.size());  
+    for(int i=0;i<probBasePoses.size();++i)
+    {
+      Eigen::VectorXd vec(7);
+      poseToEigenVector(probBasePoses[i],vec);
+      poseData.col(i)<<vec;
+    }
+    
+   
+    //Mean Centering data
+    Eigen::VectorXd featureMeans = poseData.rowwise().mean();
+    Eigen::MatrixXd centered = poseData.colwise() - featureMeans;
+    
+
+    //Computing covariance matrix
+    Eigen::MatrixXd cov = centered * centered.adjoint();
+    cov = cov/(poseData.cols());
+    
+
+    //Computing Eigen vectors and Eigen Values
+    Eigen::SelfAdjointEigenSolver<Eigen::MatrixXd> eig(cov);
+    Eigen::VectorXd eigenValues = eig.eigenvalues();
+    
+    Eigen::MatrixXd eigenVectors = eig.eigenvectors();
+    Eigen::VectorXd prncplCompnent = eigenVectors.rightCols(1);
+    
+    
+    tf2::Quaternion pc_quat(prncplCompnent[3], prncplCompnent[4], prncplCompnent[5], prncplCompnent[6]); 
```

Comment:
Did you check to see if this gives valid results?  I think that it should work, but it would be nice to validate first.  Pick two known quaternions and the correct solution (i.e. (1,0,0,0) and (0.707, 0.707,0,0), solution (0.92388, 0.38268, 0,0) and make sure that the eigen vectors return the right solution.  If the eigen vectors do not return the right solution, then you will have to do something else.  My guess is that you will have to convert the quaternions to RPY, do PCA, then convert the final solution back to a quaternion.